Integrate forked capstan (NodeJs and C/C++ only)

miha-plesko opened the pull request:
Hi @ilevine, hi @gberginc,

this PR is a response to https://github.com/emc-advanced-dev/unik/pull/118. It's focused on the very core of the forked Capstan integration only, so it's much smaller than the original one.

I'm looking forward to be hearing your opinion!

Commit:
Update documentation with forked Capstan description

Added description about how to use the new supported
languages for osv base (nodejs and native). For native (C/C++)
we assume that user has compiled her application PRIOR deploying
it into unikernel.

Signed-off-by: Miha Pleško <miha.plesko@xlab.si>

Commit:
Add container "compilers-osv-dynamic"

This container contains forked Capstan and is able
to compose arbitrary OSv image.

Signed-off-by: Miha Pleško <miha.plesko@xlab.si>

Commit:
Add compiler "OSv dynamic" (hidden from user)

OSv dynamic compiler is not exposed to the CLI directly, but
other compilers will interact with forked Capstan thru
it.

Signed-off-by: Miha Pleško <miha.plesko@xlab.si>

Commit:
Add QEMU image finisher for OSv

Images that are composed with Capstan are in QEMU QCOW2
format. Therefore we need QEMU image finisher to use
them.

Signed-off-by: Miha Pleško <miha.plesko@xlab.si>

Commit:
Support NodeJS language for OSv

Create compiler to support command:
```
unik build --base osv --provider openstack --language nodejs
```

Signed-off-by: Miha Pleško <miha.plesko@xlab.si>

Commit:
Support C/C++ language for OSv

Create compiler to support command:
```
unik build --base osv --provider openstack --language native
```

Signed-off-by: Miha Pleško <miha.plesko@xlab.si>

ilevine commented:
Hi. This pull request looks good. I will only ask that you add some documentation / markdown files explaining to users how to use the new OSv Dynamic Compiler, with an example (which would also help me to test it locally). Other than that, the code looks good and should not be a problem to merge when documentation is ready.

miha-plesko commented:
Hello, that's great news. I've added a commit that brings us some markdown files. Also, I've prepared two examples to test the new compiler (see forked-capstan-examples.zip below):

**Example-Dynamic-NodeAuto** demonstrates how to make use of OSv NodeJS compiler. It also contains a file unik_openstack_up.sh with two lines:
```
unik build --name imgDyn_node --path ./ --base osv --language nodejs --provider openstack --force
unik run --instanceName instDyn_node --imageName imgDyn_node
```

**Example-Dynamic-MySQLAuto** demonstrates how to make use of OSv native compiler (C/C++). It demonstrates how to make use of precompiled package with MySQL server.

Should you have any trouble please let me know.

[forked-capstan-examples.zip](https://github.com/emc-advanced-dev/unik/files/709329/forked-capstan-examples.zip)

gberginc commented on Makefile:
```diff
@@ -99,6 +100,7 @@ pull:
 	$(call pull_container,compilers-rump-base-hw)
 	$(call pull_container,rump-debugger-qemu)
 	$(call pull_container,compilers-rump-base-common)
+	$(call pull_container,compilers-osv-dynamic)
```
I'd prefer to keep osv compilers together and use the same order (java then dynamic)

miha-plesko replied:
Done.

gberginc commented on Makefile:
```diff
@@ -120,6 +122,7 @@ compilers: compilers-includeos-cpp-hw \
            compilers-rump-python3-hw \
            compilers-rump-python3-hw-no-stub \
            compilers-rump-python3-xen \
+           compilers-osv-dynamic \
```
Move after java.

miha-plesko replied:
Done.

gberginc commented on containers/compilers/osv/dynamic/Dockerfile:
```diff
@@ -0,0 +1,39 @@
+RUN apt-get install -y curl git qemu
+
+# Install GO
+RUN curl https://storage.googleapis.com/golang/go1.6.2.linux-amd64.tar.gz | tar xz -C /usr/local && \
```
Should we be more up to date and use the current stable Go (1.7.4)?

miha-plesko replied:
Good point, done. And tested, it works.

gberginc commented on docs/compilers/osv.md:
```diff
@@ -1,10 +1,23 @@
+either from the remote repository.
+
+Please note that UniK **does not** compile any of your application. Basically, all that it can
+do for you is to upload your files to the unikernel.
```
Please separate here. For Java, it will still compile the code.

Node and dynamic on the other hand, need to contain the files that are to be uploaded into the unikernel.

miha-plesko replied:
Done.

gberginc commented on docs/compilers/osv.md:
````diff
@@ -16,3 +29,161 @@ Compiling Java on the OSv platform requires the following parameters be met:
+config_set_default: conf1
+```
+Optionally, you can create file `manifest.yaml` directly inside your project root directory
+where you specify some additional parameters. Parameter `image_size` specifys filesystem size
````
`specifys` --> `specifies`

miha-plesko replied:
Done.

gberginc commented on docs/compilers/osv.md:
```diff
@@ -16,3 +29,161 @@ Compiling Java on the OSv platform requires the following parameters be met:
+   --path ./
+   --base osv
+   --language nodejs
+   --provider openstack
```
Will this version work with other providers, e.g. qemu or virtual box? If so, I'd suggest you replace with one of them, because it will be much easier to play with.

miha-plesko replied:
Updated to
```
--provider [qemu|openstack]
```

gberginc commented on docs/compilers/osv.md:
```diff
@@ -16,3 +29,161 @@ Compiling Java on the OSv platform requires the following parameters be met:
+b) run your own binary code
+c) combination of (a) and (b)
+In case you intend to use packages from the remote repository, please provide
+meta/run.yaml where you list desired packages.
```
Shouldn't this be `meta/package.yaml`? Furthermore, please add a reference to the additional documentation.

miha-plesko replied:
Done, had to fix native compiler actually. Thanks for noticing this.

gberginc commented on docs/compilers/osv.md:
```diff
@@ -16,3 +29,161 @@ Compiling Java on the OSv platform requires the following parameters be met:
+eu.mikelangelo-project.osv.cli
+eu.mikelangelo-project.osv.cloud-init
+eu.mikelangelo-project.osv.httpserver
+eu.mikelangelo-project.osv.nfs
```
Even though I don't like very much, I hope we can pass for the first iteration. The list of available packages could be retrieved via `capstan package search` command, but since we don't want to mess around too much with Capstan we provide the list here. However, I think we can provide an external script that will always list the current set of packages and then provide the link here.

gberginc replied:
Alternatively, we could skip this part in the README and encourage the user of unik to actually run the `describe-compiler` command where they will get up-to-data list.

miha-plesko replied:
At the moment, I'd leave it hard-coded. This PR is getting too large. But I do like the idea of external script (or maybe even already generated textfile). Perhaps we could simpy add it to the S3 repo where packages are and UniK would be able to download it and show it to the user?

Anyway, I would definetly prefer to get hard-coded values merged first, since it's becoming difficult for me to handle 7 a sequence of 7 commits in this PR...

gberginc replied:
Agree

gberginc commented on docs/compilers/osv.md:
````diff
@@ -16,3 +29,161 @@ Compiling Java on the OSv platform requires the following parameters be met:
+use. For example, if you want to run MySQL server, then the content of `meta/package.yaml`
+would be:
+```
+name: lemmy.test
````
Even though I am proud you decided to use my other nickname, I suggest to change to something else :)

miha-plesko replied:
Whooops :D

gberginc commented on pkg/compilers/osv/osv_dynamic.go:
```diff
@@ -0,0 +1,82 @@
+	"gopkg.in/yaml.v2"
+)
+
+const OSV_DEFAULT_SIZE = "10GB"
```
Is there any reason we shouldn't reduce this to 1GB? The only way I could ever go over 100MB was when I added some data to the unikernel.

miha-plesko replied:
You are correct. Fixed.

gberginc commented on pkg/compilers/osv/osv_dynamic.go:
```diff
@@ -0,0 +1,82 @@
+		return defaultMB
+	}
+	if err := yaml.Unmarshal(data, &config); err != nil {
+		return defaultMB
```
I think you should provide a warning message because you failed to get the size (and let the user know that default is being used).

miha-plesko replied:
Agreed.

gberginc commented on pkg/compilers/osv/osv_dynamic.go:
```diff
@@ -0,0 +1,82 @@
+// Returns default image size if anything goes wrong.
+func readImageSizeFromManifest(projectDir string) unikos.MegaBytes {
+	config := dynamicProjectConfig{
+		Size: OSV_DEFAULT_SIZE,
```
Just checking: in case I don't provide the `image_size` in `manifest.yaml`, this default size will still be used in line 72 below, right? Or it will return in line 68?

miha-plesko replied:
The default size will still be there in L72, yes. Parameters for Unmarshalling cannot be marked as "required" hence no error will occur if it will be missing. Furthermore, default value from L59 will not be set to empty string, so the behavior will be as expected.

gberginc replied:
Thanks for clarification!

gberginc commented on pkg/compilers/osv/osv_qemu.go:
```diff
@@ -0,0 +1,30 @@
+	ImageFinisher
+}
+
+func (b *QemuImageFinisher) FinishImage(params FinishParams) (*types.RawImage, error) {
```
Nitpicking: should this be part of a separate PR?

miha-plesko replied:
Sounds reasonable at first glance, but it's tightly coupled with other code of this PR:
* QEMU finisher uses `params.CompileParams.SizeMB` that is introduced in this PR
* this PR uses QEMU finisher in daemon.go
If possible, I'd leave this commit as part of the same PR.

gberginc replied:
Ok, this is fine with me!

gberginc commented on pkg/compilers/osv/osv_dynamic.go:
```diff
@@ -0,0 +1,82 @@
+	// append MB unit if only number was passed
+	rNoUnit, _ := regexp.Compile("^[0-9]+$")
+	if match := rNoUnit.FindStringSubmatch(config.Size); len(match) == 1 {
+		config.Size += "MB"
```
I am curious if we could just remove this and make `ParseSize` interpret number as `MB`?

miha-plesko replied:
Thanks for the suggestion. Turns out that it really is better to move logic there since it only requires additional if statement there (compared to additional import here).

gberginc reviewed (commented):
@miha-plesko I like what you did with the previous PR. I've got some additional comments, not fundamental, but hopefully may lead to some minor improvements.

Also, thanks for the two examples you provided. Do you consider these two examples could be added as references in the README?

Commit:
Add OSV examples into docs/examples

This commit adds two examples to the git, one for OSv NodeJS
and one for OSv C/C++ (MySQL) application.

Signed-off-by: Miha Pleško <miha.plesko@xlab.si>

miha-plesko commented:
Hi, I've applied the comments that @gberginc has given and provided documentation as requested by @ilevine. Please let me know if there is anything else I can do. I like to see how we're converging towards excelence :smile:

gberginc commented on Makefile:
```diff
@@ -98,7 +100,7 @@ pull:
 	$(call pull_container,compilers-rump-base-hw)
 	$(call pull_container,rump-debugger-qemu)
-	$(call pull_container,compilers-rump-base-common)
+	$(call pull_container,compilers-rump-base-common)	
```
Remove whitespace after `)`

gberginc reviewed (approved):
Just one minor whitespace to be removed, otherwise thanks for fixing and providing explanations.

@ilevine will have the final say as I can only review from a distance.
